enhancements/olm/platform-operators.md

```diff
@@ -0,0 +1,435 @@
+
+### Workflow Description
+
+```mermaid
+sequenceDiagram
+actor Admin
+participant Installer
+participant Bootstrap
+participant CVO
+participant ClusterOperator
+participant PO manager
+Installer->>Admin: Creates the Kubernetes manifests
+Installer->>PO custom resource: Generates
+Admin-->>Installer: Edits Kubernetes manifests
+Installer->>Bootstrap: Creates
+Bootstrap->>CVO: Deploys
+CVO->>PO manager: Deploys
+PO manager->>ClusterOperator: Manages
+PO manager->>PO custom resource: Watches
+CVO->>ClusterOperator: Watches
+```
+
+Note: The above workflow requires admins to manually create, and edit, the cluster Kubernetes manifests with their desired PO intent. There's a chance that admins mistype the package names, and depending on whether upfront validation is possible, that invalid configuration may not be surfaced until after cluster creation. This behavior may require explicit documentation to reduce this risk. These are obvious UX gaps with this workflow, which can be solidified in future phases, but it will result in zero installer changes during the 4.12 timeframe.
+
+### API Extensions
+
+- Adds a top-level PlatformOperator cluster-scoped CRD.
+- Adds the Bundle and BundleDeployment OLM 1.x API(s).
+- Adds a validating webhook for the Bundle and BundleDeployment APIs.
+- Adds a CRD validation webhook.
+
+### Risks and Mitigations
+
+- Early phases will introduce multiple v1alpha1 CRDs downstream.
+- Early phases will introduce OLM v1 components that are in alpha stages.
+- No concrete migration plan from legacy to OLM v1.
+- Coordination around external team dependencies.
+- The BundleDeployment API has a large net of permissions that may require upfront RBAC to limit this API to admins.
+
+### Drawbacks
+
+- Implementing this PO mechanism using legacy OLM would provide more immediate business value. That implementation is outlined below in the alternatives section.
+- Early phases will be delivered through tech preview guidelines. This leads to little customer feedback, given it precludes production clusters.
+- Early phases may introduce a new OLM 1.x control plane that has little-to-no interaction with the existing, legacy OLM 0.x control plane.
+- Early phases may introduce validating webhooks. These webhooks can impact the cluster's overall health when they're not available.
+- Early phases may introduce validating webhooks. By default, these webhooks will be deployed using an HA configuration, but may not have logic to accommodate SNO cluster environments.
+- Early phases may introduce a dynamic watching implementation for managing the underlying OLM operator resources. This has potential performance implications that need to be measured through benchmarking and/or addressed before any GA phase.
+- Future phase designs may require a large-scale rewrite of the OLM component.
+
+## Design Details
+
+### Open Questions [optional]
```

Comment:
What does migration look like for a component that's removed from the payload and becomes optional, or one that's OLM-based today and would fall under this new category?

Reply:
> What does migration look like for a component that's removed from the payload and becomes optional, 

they'd have to repackage their operator code as an OLM operator (either a registryv1 operator or a plainv1 operator), and of course they'll need some logic to remove/replace or take over the resources on the cluster that came out of the payload, so that on upgrade the PO can "takeover" the CVO-operator resources.

> or one that's OLM-based today and would fall under this new category?

one that's olm based today should potentially need no migration at all.  That said, the list of restrictions/requirements for a PO described in this doc would come into play (e.g. clusterscoped installation only).  We may also have specific channel naming requirements in order to ensure we can find the updates to the operator, but we're still working through that.

Reply:
Based on the other comment about mirroring into a disconnected environment, there would be an additional migration step for the user: expand the scope of their mirroring workflow to retrieve not just an ocp release, but also one or more olm operators. I think it's worth digging into that scenario:

- What would happen if they try to upgrade a cluster to a version where some component they're using is no longer in the payload, but that component isn't available via any OLM catalog? Can/should we block the upgrade somehow? Or would we upgrade but leave the cluster degraded?
- How can we make it easy on the user to figure out what additional things they need to mirror?

These are questions that may not arise until later when components are removed from the payload, but the packaging decisions made now will influence how that works out.

Reply:
> What would happen if they try to upgrade a cluster to a version where some component they're using is no longer in the payload, but that component isn't available via any OLM catalog? 

the upgrade should be blocked if the existing installed operator can't support the new OCP version.  If the existing installed operator can support the new OCP version, we'd allow the upgrade, you just wouldn't get a new version of that operator.

Reply:
Added this to the "Answered Questions" header.